Realm/RLMSyncPermissionResults.mm

```diff
@@ -0,0 +1,111 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2017 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import "RLMSyncPermissionResults_Private.hpp"
+
+#import "collection_notifications.hpp"
+#import "RLMSyncPermissions_Private.hpp"
+#import "RLMSyncUser_Private.hpp"
+#import "RLMUtil.hpp"
+
+using namespace realm;
+
+@interface RLMSyncPermissionResultsToken() {
+    std::unique_ptr<NotificationToken> _token;
+}
+@end
+
+@implementation RLMSyncPermissionResultsToken
+
+- (void)suppressNextNotification {
+    @throw RLMException(@"RLMPermissionResultsTokens cannot be passed to "
+                        @"the commitWriteTransactionWithoutNotifying: method");
+}
+
+- (void)stop {
+    _token = nil;
+}
+
+- (void)dealloc {
+    if (_token) {
+        NSLog(@"RLMPermissionResultsToken released without unregistering "
+              @"a notification. You must hold on to the token and call "
+              @"-[RLMPermissionResultsToken stop] when you no longer wish "
+              @"to receive notifications.");
+    }
+}
+
+- (instancetype)initWithToken:(NotificationToken)token {
+    if (self = [super init]) {
+        _token = std::make_unique<NotificationToken>(std::move(token));
+        return self;
+    }
+    return nil;
+}
+
+@end
+
+@interface RLMSyncPermissionResults () {
+    std::unique_ptr<PermissionResults> _results;
+}
+@end
+
+@implementation RLMSyncPermissionResults
+
+- (NSInteger)count {
+    REALM_ASSERT_DEBUG(_results);
+    return _results->size();
+}
+
+- (RLMSyncPermissionResultsToken *)addNotificationBlock:(RLMPermissionStatusBlock)block {
+    REALM_ASSERT_DEBUG(_results);
+    auto token = _results->async(RLMWrapPermissionStatusCallback(block));
+    return [[RLMSyncPermissionResultsToken alloc] initWithToken:std::move(token)];
+}
+
+- (RLMSyncPermissionValue *)permissionAtIndex:(NSInteger)index {
+    REALM_ASSERT_DEBUG(_results);
+    if (index < 0 || (size_t)index >= _results->size()) {
```

Comment:
`_results->get()` is already doing this check, and the exception thrown by that should just be translated to an obj-c exception both for consistency and because there are a bunch of other exceptions that could be thrown which also need to be translated.